src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        for (const [key, value] of Object.entries(extraData)) {
+          newState[key] = value;
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+    stateHistory.push(newState);
+    currentState = newState;
```

Comment:
[DONT REPEAT YOURSELF] - You are pushing `newState` to `stateHistory` and updating `currentState` in every case of the switch statement. This can be done once after the switch statement to avoid repetition and make the code cleaner.